cmds/test-dev.c

```diff
@@ -0,0 +1,303 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system loader
+ *
+ * Test device
+ *
+ * Copyright 2023 Phoenix Systems
+ * Author: Hubert Badocha
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <hal/hal.h>
+#include <lib/lib.h>
+#include <syspage.h>
+#include <devices/devs.h>
+
+#include "cmd.h"
+
+
+#define TEST_DEV_TIMEOUT_MS 500
+
+#define ERASE_BEFORE 1
+#define ERASE_AFTER  2
+
+#define BUF_SIZE 257 /* Write and read size specially not aligned to any power of 2 to possibly test more edge cases. */
+
+#define CMD_TESTDEV_READ_BIG_BUF_SIZE 0 /* Used to perform a check with a read using big buffer. */
+
+
+static void cmd_testDevInfo(void)
+{
+	lib_consolePuts("performs simple dev read/write test, usage:test-dev [-e erase before] [-E erase after] <dev> <addr> <size>");
+}
+
+
+static void cmd_testDevUsage(void)
+{
+	lib_consolePuts("Usage: test-dev [-e] [-E] <major>.<minor> <addr> <size>\n"
+					"  -e       Erase before\n"
+					"  -E 		Erase after\n");
+}
+
+
+static int cmd_testDevWrite(int major, int minor, addr_t addr, size_t length)
+{
+	ssize_t res;
+	size_t rsz, i, chunk;
+	u8 buf[BUF_SIZE];
+
+	for (rsz = 0; rsz < length; rsz += res) {
+		chunk = min((length - rsz), sizeof(buf));
+		for (i = 0; i < chunk; i++) {
+			buf[i] = ((rsz + i) & 0xffu);
+		}
+		res = devs_write(major, minor, addr + rsz, buf, chunk);
+		if (res < 0) {
+			lib_printf("\nWrite failed %zd\n", res);
+			return res;
+		}
+	}
+	return EOK;
+}
+
+
+static int cmd_testDevRead(int major, int minor, addr_t addr, size_t length)
+{
+	ssize_t res;
+	size_t rsz, chunk;
+	u8 buf[BUF_SIZE];
+
+	lib_printf("\nRead device %d.%d  from 0x%x to 0x%x (%zu bytes):\n", major, minor, (u32)addr, (u32)addr + length, length);
+	lib_consolePutHLine();
+
+	for (rsz = 0; rsz < length; rsz += res) {
+		chunk = min((length - rsz), sizeof(buf));
+		res = devs_read(major, minor, addr + rsz, buf, chunk, TEST_DEV_TIMEOUT_MS);
+		if (res < 0) {
+			log_error("\nCan't read data %zd\n", res);
+			return res;
+		}
+		lib_consolePutRegionHex((addr_t)buf, (addr_t)buf + res, addr + rsz, 0, NULL);
+	}
+	return EOK;
+}
+
+#if (CMD_TESTDEV_READ_BIG_BUF_SIZE != 0)
+
+static int cmd_testDevCheckReadBig(int major, int minor, addr_t addr, size_t length)
+{
+	static u8 buf[CMD_TESTDEV_READ_BIG_BUF_SIZE] __attribute__((section(".uncached_ddr")));
+	ssize_t res;
+	size_t rsz, i, chunk;
+
+	for (rsz = 0; rsz < length; rsz += res) {
+		chunk = min((length - rsz), sizeof(buf));
+		res = devs_read(major, minor, addr + rsz, buf, chunk, TEST_DEV_TIMEOUT_MS);
+		if (res < 0) {
+			log_error("\nCan't read data %zd\n", res);
+			return res;
+		}
+		for (i = 0; i < chunk; i++) {
+			if (buf[i] != ((rsz + i) & 0xffu)) {
+				log_error("\nDiscrepancy at index %zu, expected %zu got %u\n", rsz + i, i, buf[i]);
+				return -1;
+			}
+		}
```

Comment:
This is a good candidate for separate function - this code is duplicated in one place (two, should check be added to `cmd_testDevRead`)

Reply:
done